qiskit/circuit/library/arithmetic/weighted_adder.py

```diff
@@ -0,0 +1,324 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Compute the weighted sum of qubit states."""
+
+from typing import List, Optional
+import numpy as np
+
+from qiskit.circuit import QuantumCircuit, QuantumRegister
+
+
+class WeightedAdder(QuantumCircuit):
+    r"""A circuit to compute the weighted sum of qubit registers.
+
+    Given :math:`n` qubits :math:`q_0, \ldots, q_{n-1}` and non-negative integer weights
+    :math:`\lambda_0, \ldots, \lambda_{n-1}`, this circuit performs the operation
+
+    .. math::
+
+        |q_0 \ldots q_{n-1}\rangle |0\rangle_s
+        \mapsto |q_0 \ldots q_{n-1}\rangle |\sum_{j=0}^{n-1} \lambda_j q_j\rangle_s
```

Comment:
I think here `|q0 ... q_n-1>` means basis states, so should mention this somewhere